cli/src/commands/fix.rs

```diff
@@ -0,0 +1,328 @@
+    ///    file's path.
+    repo_path: RepoPathBuf,
+}
+
+/// Applies `run_tool()` to the inputs and stores the resulting file content.
+///
+/// Returns a map describing the subset of `tool_inputs` that resulted in
+/// changed file content. Failures when handling an input will cause it to be
+/// omitted from the return value, which is indistinguishable from succeeding
+/// with no changes.
+/// TODO: Better error handling so we can tell the user what went wrong with
+/// each failed input.
+fn fix_file_ids<'a>(
+    store: &Store,
+    tool_command: &CommandNameAndArgs,
+    tool_inputs: &'a HashSet<ToolInput>,
+) -> BackendResult<HashMap<&'a ToolInput, FileId>> {
+    let (updates_tx, updates_rx) = channel();
+    // TODO: Switch to futures, or document the decision not to. We don't need
+    // threads unless the threads will be doing more than waiting for pipes.
+    tool_inputs.into_par_iter().try_for_each_init(
+        || updates_tx.clone(),
+        |updates_tx, tool_input| -> Result<(), BackendError> {
+            let mut read = store.read_file(&tool_input.repo_path, &tool_input.file_id)?;
+            let mut old_content = vec![];
+            read.read_to_end(&mut old_content).unwrap();
+            if let Ok(new_content) = run_tool(tool_command, tool_input, &old_content) {
+                if new_content != *old_content {
+                    let new_file_id =
+                        store.write_file(&tool_input.repo_path, &mut new_content.as_slice())?;
+                    updates_tx.send((tool_input, new_file_id)).unwrap();
+                }
+            }
+            Ok(())
+        },
+    )?;
+    drop(updates_tx);
+    let mut result = HashMap::new();
+    while let Ok((tool_input, new_file_id)) = updates_rx.recv() {
+        result.insert(tool_input, new_file_id);
+    }
+    Ok(result)
+}
+
+/// Runs the `tool_command` to fix the given file content.
+///
+/// The `old_content` is assumed to be that of the `tool_input`'s `FileId`, but
+/// this is not verified.
+///
+/// Returns the new file content, whose value will be the same as `old_content`
+/// unless the command introduced changes. Returns `None` if there were any
+/// failures when starting, stopping, or communicating with the subprocess.
+fn run_tool(
+    tool_command: &CommandNameAndArgs,
+    tool_input: &ToolInput,
+    old_content: &[u8],
+) -> Result<Vec<u8>, ()> {
+    // TODO: Pipe stderr so we can tell the user which commit, file, and tool it is
+    // associated with.
+    let mut vars: HashMap<&str, &str> = HashMap::new();
+    vars.insert("path", tool_input.repo_path.as_internal_file_string());
+    let mut child = tool_command
+        .to_command_with_variables(&vars)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .or(Err(()))?;
+    let mut stdin = child.stdin.take().unwrap();
+    let mut output: Option<_> = None;
+    rayon::scope(|s| {
```

Comment:
Maybe better to use `std::thread::scope()`?

I don't know if rayon preempt sleeping threads, but if it doesn't, this might deadlock on e.g. single-core machine.

Reply:
That's a good point. I didn't consider this when I suggested rayon scope.

If I understand your point correctly, the problem is really when the rayon thread pool has 1 thread, so we only schedule 1 task concurrently (and the single-core CPU case is just an example of when we would have 1 thread), and if it blocks, then the other task can never start executing. (Alternatively, if we have n formatting jobs and each happens to block on its writing thread and none start its reading thread.)

I eventually found this note in the documentation confirming it: https://docs.rs/rayon/latest/rayon/fn.join.html#warning-about-blocking-io

(In the past, I had assigned >n blocking tasks to n threads, but IIRC they were specifically blocking on other rayon tasks, not I/O, and so could make forward progress.)

I guess it's unfortunate to spawn two new real threads every time we want to launch a subprocess, but there's not much we can easily do about it. (It looks like rayon thread pools are fixed-size, so we can't do something where we reuse an existing thread or, if no threads are currently available, instead create a new thread.)

Reply:
Btw std thread scope spawn seems to return a kind of join handle, so you could just return the data directly from the closures and `join` them after spawning both threads, without having to mutate an existing variable.

Reply:
Thanks for checking the doc. btw, we'll only need one reader or writer thread?
1. spawn writer thread
2. read and wait in the current thread
3. join at the end of the scope

Reply:
I think this actually fixed some test flakiness of the last push.